### ISSUE TITLE
Display a diff of two versions of the protocol

### ISSUE THREAD

**Issue by rom1504** [PrismarineJS/minecraft-data#136]
Would be interesting.

**Comment by rom1504** [comment]
> 16:23 `<hansihe>` i would think the optimal way to maintain the protocol files would be to start at some base version, then have each version be represented as a diff from the last
> 16:23 `<hansihe>` then generate the full files
> 16:24 `<hansihe>` keep in mind i have little knowledge on how you maintain/add new versions, but if it was done like that, changes to things like naming in one protocol version would (mostly) automatically be propagated to the others

This might be a good idea.

**Comment by rom1504** [comment]
A diff could simply contain the packet that changed.
Things that would need to be done:
* [ ] script to generate the diff between 2 protocol.json
* [ ] script to generate the full file from a base and a diff

**Comment by roblabla** [comment]
Lol. This was the way I originally wanted to handle the protocol data, and after a long argument with... erm, either myself or you " ? I opted to go down the "full protocol.json route".

A pro of diffs is that if there's a bug in 1.5 that snowballs to 1.6, 1.7, etc..., fixing it in 1.5 will fix all affected versions. The con is obviously that if 1.6 didn't have that change, well now you broke the later versions.

Another con is that each protocol.json becomes dependant on the previous ones to make sense.

**Comment by rom1504** [comment]
Yeah we argued about that a long time ago. My problem with it (as Gjum just said on gitter) was mostly that then users have to generate the protocol.json themselves.
But if we have the diff + the full file then it's convenient enough.
(we change the diff manually, and generate the full file automatically)

**Comment by rom1504** [comment]
And it makes much more sense now that we support snapshots and many small versions, there are little diff between those. And that the name-id mapping is just in one packet.

**Comment by roblabla** [comment]
just a quick idea, we can have a git hook that autogenerates the full file from the diffs pretty easily.
so we don't forget. Also will want to have the build status make sure the generated file matches the in-tree file

**Comment by rom1504** [comment]
Here is a proposal for a json diff format:
* [from 16w20a to 1.10-pre1](https://gist.github.com/rom1504/a14387eaf2bfb78698bc9d8dfcd4cbc6)
* [from 1.8 to 1.9.4](https://gist.github.com/rom1504/7303d8de115af0db8caee728f3d37f6f)
* [the script](https://gist.github.com/rom1504/d7bc33522c071a943fad4833fd1e5dd8)

**Comment by roblabla** [comment]
Why the need to specify the oldType ? Is it purely for diff readability ?

**Comment by rom1504** [comment]
Because without it we don't know what packets to remove. Also it makes the diff reversable (which might be useful I guess ?)

**Comment by rom1504** [comment]
Although if we don't care about reversability, we could add a "removedTypes" with just a list of names.

**Comment by phase** [comment]
I think the full jsons are better for projects that will only be using the information of one protocol. I don't want to have to patch a base json with a diff if I can use a complete one instead.

**Comment by rom1504** [comment]
Even if we put json diff in the repo, the full files will still be available ;)

**Comment by roblabla** [comment]
@phase the idea is to make maintenance (AKA editing the files) easier, along with making the differences easier to see. The full file will still be available, but will be auto-generated when pushing a modifications by going from the base file and applying the patch files.
